# Stage 3 - Modeling Next-Night Sleep Outcomes


## Framing

This notebook keeps the original `day D -> next night` contract, but stops assuming that one binary cutoff (`recovery < 70`) is the only sensible modeling target.

What this notebook now does:
- screens several recovery thresholds (`70`, `75`, `80`, and the empirical median) for the binary classification task
- keeps the main classification story tied to next-night recovery risk
- compares several continuous next-night targets with different distributions: `sleepRecoveryScore`, `sleepOverallScore`, `sleepQualityScore`, and `avgSleepStress`

The goal is not to force one device metric to work. The goal is to see which target/task pair produces the clearest and most defensible Stage 3 signal.


In [1]:
from pathlib import Path
import sys

import pandas as pd
from IPython.display import display

repo_root = Path.cwd().resolve()
if not (repo_root / 'src').exists():
    repo_root = repo_root.parent
if str(repo_root / 'src') not in sys.path:
    sys.path.insert(0, str(repo_root / 'src'))

from garmin_analytics.modeling.prepare import (
    DEFAULT_LOW_RECOVERY_THRESHOLD,
    load_recovery_modeling_frame,
    resolve_recovery_feature_columns,
)
from garmin_analytics.modeling.recovery import (
    build_confusion_table,
    build_feature_importance_table,
    build_permutation_importance_table,
    evaluate_recovery_classification_models_tuned,
    evaluate_recovery_regression_models,
    forward_select_classification_features,
)

pd.set_option('display.max_columns', 200)
pd.set_option('display.width', 160)


In [2]:
daily_path = repo_root / 'data' / 'processed' / 'daily_sanitized.parquet'
quality_path = repo_root / 'data' / 'processed' / 'daily_quality.parquet'
DEFAULT_RECOVERY_THRESHOLD = DEFAULT_LOW_RECOVERY_THRESHOLD
RECOVERY_SCREEN_THRESHOLDS = [70.0, 75.0, 80.0]
PRIMARY_RECOVERY_THRESHOLD = 75.0

print('repo_root:', repo_root)
print('daily_path exists:', daily_path.exists(), daily_path)
print('quality_path exists:', quality_path.exists(), quality_path)
print('default recovery threshold:', DEFAULT_RECOVERY_THRESHOLD)
print('recovery thresholds to screen:', RECOVERY_SCREEN_THRESHOLDS)
print('primary recovery threshold for deep-dive:', PRIMARY_RECOVERY_THRESHOLD)


repo_root: /Users/abatrakov/Documents/FUN/wearable-analytics
daily_path exists: True /Users/abatrakov/Documents/FUN/wearable-analytics/data/processed/daily_sanitized.parquet
quality_path exists: True /Users/abatrakov/Documents/FUN/wearable-analytics/data/processed/daily_quality.parquet
default recovery threshold: 70.0
recovery thresholds to screen: [70.0, 75.0, 80.0]
primary recovery threshold for deep-dive: 75.0


In [3]:
model_df = load_recovery_modeling_frame(
    daily_path,
    quality_path,
    current_day_quality='strict',
    low_recovery_threshold=DEFAULT_RECOVERY_THRESHOLD,
    include_schedule_targets=True,
)

recovery_score = pd.to_numeric(model_df['target_sleepRecoveryScore_next_night'], errors='coerce')
recovery_median = float(recovery_score.dropna().median())
recovery_thresholds = []
seen_thresholds = set()
for threshold in [*RECOVERY_SCREEN_THRESHOLDS, recovery_median]:
    threshold = float(threshold)
    if threshold not in seen_thresholds:
        recovery_thresholds.append(threshold)
        seen_thresholds.add(threshold)

summary = pd.DataFrame(
    [
        {
            'rows_total': len(model_df),
            'rows_with_recovery_target': int(model_df['target_sleepRecoveryScore_next_night'].notna().sum()),
            'rows_with_overall_target': int(model_df['target_sleepOverallScore_next_night'].notna().sum()),
            'rows_with_quality_target': int(model_df['target_sleepQualityScore_next_night'].notna().sum()),
            'rows_with_avg_sleep_stress_target': int(model_df['target_avgSleepStress_next_night'].notna().sum()),
            'recovery_median': recovery_median,
        }
    ]
)

recovery_threshold_summary = pd.DataFrame(
    [
        {
            'threshold': threshold,
            'positive_rate_lt_threshold': float((recovery_score < threshold).mean()),
        }
        for threshold in recovery_thresholds
    ]
)

sample_cols = [
    'calendarDate',
    'target_sleepRecoveryScore_next_night',
    'target_sleepOverallScore_next_night',
    'target_sleepQualityScore_next_night',
    'target_avgSleepStress_next_night',
    'awakeAverageStressLevel',
    'bodyBatteryLowest',
    'active_hours',
    'nextsleep_sleep_start_hour_local_wrapped',
    'nextsleep_sleep_opportunity_hours',
]

available_sample_cols = [col for col in sample_cols if col in model_df.columns]

display(summary)
display(recovery_threshold_summary)
display(model_df[available_sample_cols].head(12))


,rows_total,rows_with_recovery_target,rows_with_overall_target,rows_with_quality_target,rows_with_avg_sleep_stress_target,recovery_median
0,617,543,543,543,544,78.0


,threshold,positive_rate_lt_threshold
0,70.0,0.209945
1,75.0,0.349908
2,80.0,0.526703
3,78.0,0.447514


,calendarDate,target_sleepRecoveryScore_next_night,target_sleepOverallScore_next_night,target_sleepQualityScore_next_night,target_avgSleepStress_next_night,awakeAverageStressLevel,bodyBatteryLowest,active_hours,nextsleep_sleep_start_hour_local_wrapped,nextsleep_sleep_opportunity_hours
0,2023-05-26,100,98,97,5.48,50,36,0.356944,2.800000,7.9
1,2023-05-27,100,82,88,8.59,41,22,1.7925,3.383333,7.549722
2,2023-05-28,100,90,89,8.02,58,8,0.366944,2.583333,8.816667
3,2023-05-29,100,85,93,6.59,36,18,1.527222,2.950000,7.133333
4,2023-05-30,87,69,75,13.16,38,24,0.936667,1.250000,11.532778
5,2023-05-31,100,87,85,7.44,46,22,0.638889,1.883333,8.066667
6,2023-06-01,100,67,93,8.79,51,20,1.285278,2.183333,5.25
7,2023-06-02,100,73,93,8.94,42,12,1.223611,6.933333,5.766667
8,2023-06-03,100,93,93,7.58,40,6,2.181667,4.350000,9.733333
9,2023-06-04,100,96,95,3.55,70,5,0.668056,1.883333,8.433333


## Feature sets


In [4]:
awake_only_cols = resolve_recovery_feature_columns(
    model_df,
    tiers=('tier1',),
    include_schedule=False,
)
awake_plus_compact_stress_cols = resolve_recovery_feature_columns(
    model_df,
    tiers=('tier1', 'tier2_compact'),
    include_schedule=False,
)
awake_plus_stress_cols = resolve_recovery_feature_columns(
    model_df,
    tiers=('tier1', 'tier2'),
    include_schedule=False,
)
awake_plus_compact_stress_plus_schedule_cols = resolve_recovery_feature_columns(
    model_df,
    tiers=('tier1', 'tier2_compact'),
    include_schedule=True,
)

feature_sets = {
    'awake_only': awake_only_cols,
    'awake_plus_compact_stress': awake_plus_compact_stress_cols,
    'awake_plus_stress': awake_plus_stress_cols,
    'awake_plus_compact_stress_plus_schedule': awake_plus_compact_stress_plus_schedule_cols,
}

screening_feature_sets = {
    name: cols
    for name, cols in feature_sets.items()
    if name != 'awake_plus_compact_stress_plus_schedule'
}

display(
    pd.DataFrame(
        [
            {
                'feature_set': name,
                'feature_count': len(cols),
                'sample_features': ', '.join(cols[:10]),
            }
            for name, cols in feature_sets.items()
        ]
    )
)


,feature_set,feature_count,sample_features
0,awake_only,21,"totalSteps, totalDistanceMeters, activeKilocal..."
1,awake_plus_compact_stress,26,"totalSteps, totalDistanceMeters, activeKilocal..."
2,awake_plus_stress,45,"totalSteps, totalDistanceMeters, activeKilocal..."
3,awake_plus_compact_stress_plus_schedule,28,"totalSteps, totalDistanceMeters, activeKilocal..."


## Recovery threshold screening


In [5]:
def add_threshold_target(frame: pd.DataFrame, *, source_col: str, threshold: float, target_col: str) -> pd.DataFrame:
    out = frame.copy()
    score = pd.to_numeric(out[source_col], errors='coerce')
    out[target_col] = score.lt(threshold).where(score.notna(), pd.NA).astype('Int64')
    return out

recovery_threshold_results = []

for threshold in recovery_thresholds:
    target_col = f'target_low_recovery_lt_{str(threshold).replace(".", "_")}'
    threshold_df = add_threshold_target(
        model_df,
        source_col='target_sleepRecoveryScore_next_night',
        threshold=threshold,
        target_col=target_col,
    )
    for feature_set_name, cols in screening_feature_sets.items():
        out, _ = evaluate_recovery_classification_models_tuned(
            threshold_df,
            feature_cols=cols,
            target_col=target_col,
            threshold_metric='balanced_accuracy',
        )
        out['feature_set'] = feature_set_name
        out['recovery_threshold'] = threshold
        recovery_threshold_results.append(out)

recovery_threshold_results = pd.concat(recovery_threshold_results, ignore_index=True)
recovery_threshold_test = recovery_threshold_results[recovery_threshold_results['split'] == 'test'].copy()
recovery_threshold_test = recovery_threshold_test.sort_values(
    ['recovery_threshold', 'balanced_accuracy', 'roc_auc', 'pr_auc'],
    ascending=[True, False, False, False],
)

best_by_threshold = recovery_threshold_test.drop_duplicates(['recovery_threshold'], keep='first')

display(recovery_threshold_test)
display(best_by_threshold)


,model,split,n_rows,feature_count,threshold,threshold_metric,valid_metric_at_threshold,accuracy,balanced_accuracy,f1,positive_rate_true,positive_rate_pred,roc_auc,pr_auc,feature_set,recovery_threshold
29,logistic_regression_elasticnet,test,109,26,0.55,balanced_accuracy,0.691392,0.559633,0.633995,0.510204,0.284404,0.614679,0.591811,0.319198,awake_plus_compact_stress,70.0
47,logistic_regression_elasticnet,test,109,45,0.50,balanced_accuracy,0.647741,0.522936,0.627792,0.509434,0.284404,0.688073,0.586849,0.322337,awake_plus_stress,70.0
44,logistic_regression_l1,test,109,45,0.55,balanced_accuracy,0.674908,0.541284,0.611456,0.489796,0.284404,0.614679,0.582713,0.316237,awake_plus_stress,70.0
32,random_forest,test,109,26,0.30,balanced_accuracy,0.746642,0.522936,0.608354,0.490196,0.284404,0.651376,0.630687,0.418221,awake_plus_compact_stress,70.0
23,logistic_regression,test,109,26,0.65,balanced_accuracy,0.663004,0.577982,0.598222,0.465116,0.284404,0.504587,0.611249,0.339852,awake_plus_compact_stress,70.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
110,dummy_most_frequent,test,109,21,0.50,balanced_accuracy,0.500000,0.577982,0.500000,0.732558,0.577982,1.000000,0.500000,0.577982,awake_only,80.0
128,dummy_most_frequent,test,109,26,0.50,balanced_accuracy,0.500000,0.577982,0.500000,0.732558,0.577982,1.000000,0.500000,0.577982,awake_plus_compact_stress,80.0
146,dummy_most_frequent,test,109,45,0.50,balanced_accuracy,0.500000,0.577982,0.500000,0.732558,0.577982,1.000000,0.500000,0.577982,awake_plus_stress,80.0
119,logistic_regression_elasticnet,test,109,21,0.50,balanced_accuracy,0.588715,0.541284,0.494651,0.666667,0.577982,0.798165,0.622843,0.731359,awake_only,80.0


,model,split,n_rows,feature_count,threshold,threshold_metric,valid_metric_at_threshold,accuracy,balanced_accuracy,f1,positive_rate_true,positive_rate_pred,roc_auc,pr_auc,feature_set,recovery_threshold
29,logistic_regression_elasticnet,test,109,26,0.55,balanced_accuracy,0.691392,0.559633,0.633995,0.510204,0.284404,0.614679,0.591811,0.319198,awake_plus_compact_stress,70.0
89,hist_gradient_boosting,test,109,26,0.35,balanced_accuracy,0.612373,0.605505,0.639126,0.605505,0.385321,0.614679,0.642502,0.471894,awake_plus_compact_stress,75.0
215,hist_gradient_boosting,test,109,45,0.50,balanced_accuracy,0.629911,0.642202,0.631356,0.697674,0.541284,0.642202,0.644407,0.656802,awake_plus_stress,78.0
149,logistic_regression,test,109,45,0.60,balanced_accuracy,0.643750,0.678899,0.684092,0.700855,0.577982,0.495413,0.724638,0.789495,awake_plus_stress,80.0


## Selected recovery threshold deep-dive


In [6]:
selected_target_col = f'target_low_recovery_lt_{str(PRIMARY_RECOVERY_THRESHOLD).replace(".", "_")}'
model_df_selected = add_threshold_target(
    model_df,
    source_col='target_sleepRecoveryScore_next_night',
    threshold=PRIMARY_RECOVERY_THRESHOLD,
    target_col=selected_target_col,
)

classification_tuned_results = []
classification_fitted_by_feature_set = {}

for feature_set_name, cols in feature_sets.items():
    out, fitted = evaluate_recovery_classification_models_tuned(
        model_df_selected,
        feature_cols=cols,
        target_col=selected_target_col,
        threshold_metric='balanced_accuracy',
    )
    out['feature_set'] = feature_set_name
    out['recovery_threshold'] = PRIMARY_RECOVERY_THRESHOLD
    classification_tuned_results.append(out)
    classification_fitted_by_feature_set[feature_set_name] = fitted

classification_tuned_results = pd.concat(classification_tuned_results, ignore_index=True)
display(
    classification_tuned_results[classification_tuned_results['split'] == 'test']
    .sort_values(['balanced_accuracy', 'roc_auc', 'pr_auc'], ascending=[False, False, False])
)


,model,split,n_rows,feature_count,threshold,threshold_metric,valid_metric_at_threshold,accuracy,balanced_accuracy,f1,positive_rate_true,positive_rate_pred,roc_auc,pr_auc,feature_set,recovery_threshold
35,hist_gradient_boosting,test,109,26,0.35,balanced_accuracy,0.612373,0.605505,0.639126,0.605505,0.385321,0.614679,0.642502,0.471894,awake_plus_compact_stress,75.0
71,hist_gradient_boosting,test,109,28,0.30,balanced_accuracy,0.625579,0.577982,0.625622,0.603448,0.385321,0.678899,0.657072,0.494654,awake_plus_compact_stress_plus_schedule,75.0
53,hist_gradient_boosting,test,109,45,0.35,balanced_accuracy,0.682113,0.550459,0.589908,0.566372,0.385321,0.651376,0.647477,0.501191,awake_plus_stress,75.0
32,random_forest,test,109,26,0.40,balanced_accuracy,0.609824,0.513761,0.582267,0.582677,0.385321,0.779817,0.618337,0.449153,awake_plus_compact_stress,75.0
68,random_forest,test,109,28,0.40,balanced_accuracy,0.629055,0.504587,0.574805,0.578125,0.385321,0.788991,0.583511,0.429889,awake_plus_compact_stress_plus_schedule,75.0
26,logistic_regression_l1,test,109,26,0.65,balanced_accuracy,0.658248,0.568807,0.564854,0.494624,0.385321,0.467890,0.584577,0.433584,awake_plus_compact_stress,75.0
62,logistic_regression_l1,test,109,28,0.65,balanced_accuracy,0.658248,0.568807,0.564854,0.494624,0.385321,0.467890,0.584577,0.433584,awake_plus_compact_stress_plus_schedule,75.0
41,logistic_regression,test,109,45,0.60,balanced_accuracy,0.626738,0.522936,0.563077,0.543860,0.385321,0.660550,0.545487,0.464286,awake_plus_stress,75.0
65,logistic_regression_elasticnet,test,109,28,0.65,balanced_accuracy,0.659407,0.541284,0.551350,0.500000,0.385321,0.532110,0.557569,0.421598,awake_plus_compact_stress_plus_schedule,75.0
29,logistic_regression_elasticnet,test,109,26,0.65,balanced_accuracy,0.653383,0.541284,0.551350,0.500000,0.385321,0.532110,0.553660,0.417975,awake_plus_compact_stress,75.0


## Automated feature selection


In [7]:
selection_history_logistic, selected_features_logistic = forward_select_classification_features(
    model_df_selected,
    candidate_feature_cols=feature_sets['awake_plus_compact_stress'],
    target_col=selected_target_col,
    model_name='logistic_regression',
    selection_metric='balanced_accuracy',
    threshold_metric='balanced_accuracy',
    max_features=8,
    min_improvement=0.01,
)

selection_history_l1, selected_features_l1 = forward_select_classification_features(
    model_df_selected,
    candidate_feature_cols=feature_sets['awake_plus_compact_stress'],
    target_col=selected_target_col,
    model_name='logistic_regression_l1',
    selection_metric='balanced_accuracy',
    threshold_metric='balanced_accuracy',
    max_features=8,
    min_improvement=0.01,
)

selected_feature_sets = {
    'forward_selected_logistic': selected_features_logistic,
    'forward_selected_logistic_l1': selected_features_l1,
}

selected_feature_set_results = []
selected_feature_set_fitted = {}
for feature_set_name, cols in selected_feature_sets.items():
    out, fitted = evaluate_recovery_classification_models_tuned(
        model_df_selected,
        feature_cols=cols,
        target_col=selected_target_col,
        threshold_metric='balanced_accuracy',
        model_names=['logistic_regression', 'logistic_regression_l1', 'hist_gradient_boosting', 'random_forest'],
    )
    out['feature_set'] = feature_set_name
    selected_feature_set_results.append(out)
    selected_feature_set_fitted[feature_set_name] = fitted

selected_feature_set_results = pd.concat(selected_feature_set_results, ignore_index=True)

display(selection_history_logistic)
display(selection_history_l1)
display(
    pd.DataFrame(
        [
            {'feature_set': name, 'feature_count': len(cols), 'features': ', '.join(cols)}
            for name, cols in selected_feature_sets.items()
        ]
    )
)
display(
    selected_feature_set_results[selected_feature_set_results['split'] == 'test']
    .sort_values(['balanced_accuracy', 'roc_auc', 'pr_auc'], ascending=[False, False, False])
)


,step,model,selection_metric,added_feature,n_selected_features,selected_features,candidate_score,improvement_from_prev,threshold,valid_balanced_accuracy,valid_f1,valid_roc_auc,valid_pr_auc,valid_positive_rate_pred
0,1,logistic_regression,balanced_accuracy,awakeHighStressShare,1,awakeHighStressShare,0.652224,NaN,0.55,0.652224,0.472222,0.677479,0.399946,0.422018
1,2,logistic_regression,balanced_accuracy,dayofweek,2,"awakeHighStressShare, dayofweek",0.691844,0.039620,0.55,0.691844,0.512821,0.668211,0.374733,0.477064
2,3,logistic_regression,balanced_accuracy,activeSeconds,3,"awakeHighStressShare, dayofweek, activeSeconds",0.717099,0.025255,0.55,0.717099,0.538462,0.677943,0.373097,0.477064


,step,model,selection_metric,added_feature,n_selected_features,selected_features,candidate_score,improvement_from_prev,threshold,valid_balanced_accuracy,valid_f1,valid_roc_auc,valid_pr_auc,valid_positive_rate_pred
0,1,logistic_regression_l1,balanced_accuracy,awakeAverageStressLevel,1,awakeAverageStressLevel,0.661724,NaN,0.55,0.661724,0.481928,0.677016,0.358688,0.522936
1,2,logistic_regression_l1,balanced_accuracy,restingHeartRate,2,"awakeAverageStressLevel, restingHeartRate",0.681186,0.019462,0.60,0.681186,0.507937,0.684198,0.416657,0.339450


,feature_set,feature_count,features
0,forward_selected_logistic,3,"awakeHighStressShare, dayofweek, activeSeconds"
1,forward_selected_logistic_l1,2,"awakeAverageStressLevel, restingHeartRate"


,model,split,n_rows,feature_count,threshold,threshold_metric,valid_metric_at_threshold,accuracy,balanced_accuracy,f1,positive_rate_true,positive_rate_pred,roc_auc,pr_auc,feature_set
17,logistic_regression_l1,test,109,2,0.60,balanced_accuracy,0.681186,0.678899,0.681059,0.623656,0.385321,0.467890,0.714108,0.603263,forward_selected_logistic_l1
14,logistic_regression,test,109,2,0.60,balanced_accuracy,0.675162,0.669725,0.673596,0.617021,0.385321,0.477064,0.717662,0.607125,forward_selected_logistic_l1
20,hist_gradient_boosting,test,109,2,0.50,balanced_accuracy,0.612836,0.660550,0.630597,0.531646,0.385321,0.339450,0.614428,0.511011,forward_selected_logistic_l1
23,random_forest,test,109,2,0.55,balanced_accuracy,0.564643,0.623853,0.582978,0.453333,0.385321,0.302752,0.568941,0.467835,forward_selected_logistic_l1
8,hist_gradient_boosting,test,109,3,0.45,balanced_accuracy,0.652224,0.577982,0.581201,0.520833,0.385321,0.495413,0.655828,0.585484,forward_selected_logistic
2,logistic_regression,test,109,3,0.55,balanced_accuracy,0.717099,0.532110,0.557214,0.523364,0.385321,0.596330,0.551528,0.405939,forward_selected_logistic
5,logistic_regression_l1,test,109,3,0.55,balanced_accuracy,0.703892,0.522936,0.545309,0.509434,0.385321,0.587156,0.559346,0.410666,forward_selected_logistic
11,random_forest,test,109,3,0.45,balanced_accuracy,0.631603,0.513761,0.542289,0.513761,0.385321,0.614679,0.637527,0.524446,forward_selected_logistic


## Main candidate: sparse logistic plus nonlinear benchmark


In [8]:
main_feature_set = 'forward_selected_logistic_l1'
main_model_name = 'logistic_regression_l1'

main_test_row = (
    selected_feature_set_results[
        (selected_feature_set_results['feature_set'] == main_feature_set)
        & (selected_feature_set_results['model'] == main_model_name)
        & (selected_feature_set_results['split'] == 'test')
    ]
    .iloc[0]
)
main_threshold = float(main_test_row['threshold'])
main_feature_cols = selected_feature_sets[main_feature_set]
main_pipeline = selected_feature_set_fitted[main_feature_set][main_model_name]

frame_for_split = model_df_selected[['calendarDate', selected_target_col, *main_feature_cols]].copy()
frame_for_split = frame_for_split.dropna(subset=[selected_target_col])
frame_for_split[selected_target_col] = pd.to_numeric(frame_for_split[selected_target_col], errors='coerce').astype(int)
frame_for_split['calendarDate'] = pd.to_datetime(frame_for_split['calendarDate'], errors='coerce')
frame_for_split = frame_for_split.sort_values('calendarDate').reset_index(drop=True)

n_rows = len(frame_for_split)
train_end = max(1, int(n_rows * 0.6))
valid_end = max(train_end + 1, int(n_rows * 0.8))
valid_end = min(valid_end, n_rows - 1)
test_df = frame_for_split.iloc[valid_end:].copy()

test_proba = main_pipeline.predict_proba(test_df[main_feature_cols])[:, 1]
confusion_tbl = build_confusion_table(
    test_df[selected_target_col],
    test_proba,
    threshold=main_threshold,
)

display(main_test_row.to_frame().T)
display(confusion_tbl)

logistic_pipeline = selected_feature_set_fitted['forward_selected_logistic_l1']['logistic_regression_l1']
logistic_importance_tbl = build_feature_importance_table(logistic_pipeline, top_n=12)
print('Feature importance for forward_selected_logistic_l1 / logistic_regression_l1')
display(logistic_importance_tbl)

hgb_feature_set = 'forward_selected_logistic'
hgb_model_name = 'hist_gradient_boosting'
hgb_feature_cols = selected_feature_sets[hgb_feature_set]
hgb_pipeline = selected_feature_set_fitted[hgb_feature_set][hgb_model_name]

hgb_frame_for_split = model_df_selected[['calendarDate', selected_target_col, *hgb_feature_cols]].copy()
hgb_frame_for_split = hgb_frame_for_split.dropna(subset=[selected_target_col])
hgb_frame_for_split[selected_target_col] = pd.to_numeric(hgb_frame_for_split[selected_target_col], errors='coerce').astype(int)
hgb_frame_for_split['calendarDate'] = pd.to_datetime(hgb_frame_for_split['calendarDate'], errors='coerce')
hgb_frame_for_split = hgb_frame_for_split.sort_values('calendarDate').reset_index(drop=True)

hgb_n_rows = len(hgb_frame_for_split)
hgb_train_end = max(1, int(hgb_n_rows * 0.6))
hgb_valid_end = max(hgb_train_end + 1, int(hgb_n_rows * 0.8))
hgb_valid_end = min(hgb_valid_end, hgb_n_rows - 1)
hgb_test_df = hgb_frame_for_split.iloc[hgb_valid_end:].copy()

hgb_importance_tbl = build_permutation_importance_table(
    hgb_pipeline,
    hgb_test_df[hgb_feature_cols],
    hgb_test_df[selected_target_col],
    top_n=12,
    scoring='balanced_accuracy',
    n_repeats=20,
)
print('Permutation importance for forward_selected_logistic / hist_gradient_boosting')
display(hgb_importance_tbl)


,model,split,n_rows,feature_count,threshold,threshold_metric,valid_metric_at_threshold,accuracy,balanced_accuracy,f1,positive_rate_true,positive_rate_pred,roc_auc,pr_auc,feature_set
17,logistic_regression_l1,test,109,2,0.6,balanced_accuracy,0.681186,0.678899,0.681059,0.623656,0.385321,0.46789,0.714108,0.603263,forward_selected_logistic_l1


,threshold,tn,fp,fn,tp,precision,recall,specificity
0,0.6,45,22,13,29,0.568627,0.690476,0.671642


Feature importance for forward_selected_logistic_l1 / logistic_regression_l1


,feature,coefficient,importance_abs,direction
0,awakeAverageStressLevel,0.383682,0.383682,positive
1,restingHeartRate,0.116340,0.116340,positive


Permutation importance for forward_selected_logistic / hist_gradient_boosting


,feature,importance_mean,importance_std
0,awakeHighStressShare,0.051990,0.042038
1,dayofweek,0.030775,0.020656
2,activeSeconds,0.015965,0.024680


## Alternative regression targets


In [9]:
regression_feature_sets = {
    'awake_only': feature_sets['awake_only'],
    'awake_plus_compact_stress': feature_sets['awake_plus_compact_stress'],
}

regression_target_specs = [
    {
        'target_col': 'target_sleepRecoveryScore_next_night',
        'target_label': 'sleepRecoveryScore',
        'target_bounds': (0.0, 100.0),
    },
    {
        'target_col': 'target_sleepOverallScore_next_night',
        'target_label': 'sleepOverallScore',
        'target_bounds': (0.0, 100.0),
    },
    {
        'target_col': 'target_sleepQualityScore_next_night',
        'target_label': 'sleepQualityScore',
        'target_bounds': (0.0, 100.0),
    },
    {
        'target_col': 'target_avgSleepStress_next_night',
        'target_label': 'avgSleepStress',
        'target_bounds': None,
    },
]

regression_results = []
for spec in regression_target_specs:
    for feature_set_name, cols in regression_feature_sets.items():
        out = evaluate_recovery_regression_models(
            model_df,
            feature_cols=cols,
            target_col=spec['target_col'],
            target_bounds=spec['target_bounds'],
        )
        out['feature_set'] = feature_set_name
        out['target_label'] = spec['target_label']
        regression_results.append(out)

regression_results = pd.concat(regression_results, ignore_index=True)
regression_test = regression_results[regression_results['split'] == 'test'].copy()
regression_test = regression_test.sort_values(['target_label', 'r2', 'mae'], ascending=[True, False, True])

best_regression_by_target = regression_test.drop_duplicates(['target_label'], keep='first')

display(regression_test)
display(best_regression_by_target)


,model,split,n_rows,feature_count,target_bounds,mae,rmse,r2,pred_min,pred_max,feature_set,target_label
140,random_forest,test,109,26,none,5.861894,8.836052,0.130566,12.615353,29.897263,awake_plus_compact_stress,avgSleepStress
143,hist_gradient_boosting,test,109,26,none,5.936799,9.172811,0.063031,11.015305,30.712068,awake_plus_compact_stress,avgSleepStress
134,lasso,test,109,26,none,6.453357,9.299560,0.036958,10.468995,27.771409,awake_plus_compact_stress,avgSleepStress
137,elastic_net,test,109,26,none,6.450501,9.317709,0.033196,10.660273,28.053590,awake_plus_compact_stress,avgSleepStress
131,ridge,test,109,26,none,6.550264,9.364193,0.023525,10.230433,28.685420,awake_plus_compact_stress,avgSleepStress
119,elastic_net,test,109,21,none,5.928509,9.400377,0.015964,12.552232,22.664585,awake_only,avgSleepStress
116,lasso,test,109,21,none,5.925684,9.416111,0.012668,12.689935,22.629679,awake_only,avgSleepStress
113,ridge,test,109,21,none,5.992941,9.447193,0.006138,11.831101,22.917359,awake_only,avgSleepStress
122,random_forest,test,109,21,none,5.670333,9.508928,-0.006893,11.895546,22.356179,awake_only,avgSleepStress
110,dummy_median,test,109,21,none,5.790688,9.702729,-0.048354,15.855000,15.855000,awake_only,avgSleepStress


,model,split,n_rows,feature_count,target_bounds,mae,rmse,r2,pred_min,pred_max,feature_set,target_label
140,random_forest,test,109,26,none,5.861894,8.836052,0.130566,12.615353,29.897263,awake_plus_compact_stress,avgSleepStress
38,dummy_median,test,109,21,"(0.0, 100.0)",9.357798,13.694571,-0.010201,82.000000,82.000000,awake_only,sleepOverallScore
74,dummy_median,test,109,21,"(0.0, 100.0)",8.339450,12.695083,-0.000502,83.000000,83.000000,awake_only,sleepQualityScore
32,random_forest,test,109,26,"(0.0, 100.0)",16.226500,22.935483,0.098579,38.678336,89.981465,awake_plus_compact_stress,sleepRecoveryScore


## Interpretation summary

The refreshed run keeps `next-night sleepRecoveryScore < 75` as the main Stage 3 classification target. The modeling frame now has **617** strict-quality rows, with **543** rows carrying next-night recovery / overall / quality targets and **544** rows carrying next-night `avgSleepStress`.

Current refreshed results:

- `recovery < 75` remains the best practical cutoff: it is adverse enough to mean something, but not as sparse as `<70`.
- The best public classifier is the forward-selected sparse L1 logistic model using `awakeAverageStressLevel` and `restingHeartRate`. On the 109-row test split it reaches balanced accuracy ≈0.681, ROC-AUC ≈0.714, PR-AUC ≈0.603, and F1 ≈0.624.
- A second small logistic variant using `awakeHighStressShare`, `dayofweek`, and `activeSeconds` is directionally similar but slightly weaker on the headline test metrics.
- Nonlinear benchmarks remain useful controls, but they do not beat the selected sparse classifier in the refreshed primary story.
- Numeric regression remains limited: `avgSleepStress` and `sleepRecoveryScore` show only small positive test `R^2`, while exact next-night score prediction is still not reliable enough to replace the binary risk framing.

The honest takeaway is that daytime wearable summaries can support a moderate next-night low-recovery risk flag, especially through awake stress and resting-heart-rate context, but the current aggregate feature layer is not precise enough for decision-grade score prediction.
